### PR TITLE
Pod Level authorization: add odahu-flow-api authz policies into helm …

### LINKED ISSUE #408: Migrate to OPA sidecars and pod-specific policies
#### Brief

We need a way to assign different OPA policies per POD / APP

#### Current state:

1. we use centralized OPA deployment that makes authorization decisions on requests to each of our protected pods.
https://docs.odahu.org/gen_security.html#api-request-lifecycle
2. In policies we rely on a request context in general. But this context contains not enough information about the pod/service to which this policy is enforced so we have not permanent pod IP address as an authority and so on
Please looks at an example of an authorization request
```
{
    "destination": {
      "address": {
        "Address": {
          "SocketAddress": {
            "PortSpecifier": {
              "PortValue": 8012
            },
            "address": "10.44.9.3"
          }
        }
      }
    },
    "request": {
      "http": {
        "headers": {
          ":authority": "10.44.9.3:8012",
          ":method": "GET",
          ":path": "/api/model/info",
          "accept": "application/json",
          "accept-encoding": "gzip, deflate, br",
          "accept-language": "en-US,en;q=0.9,ru;q=0.8",
          "authorization": "",
          "cookie": "_oauth2_proxy_1=",
          "k-original-host": "gke01.dev.odahu.org",
          "k-proxy-request": "activator",
          "referer": "https://gke01.dev.odahu.org/service-catalog/swagger/index.html",
          "sec-fetch-dest": "empty",
          "sec-fetch-mode": "cors",
          "sec-fetch-site": "same-origin",
          "user-agent": "Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_5) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/85.0.4183.121 Safari/537.36",
          "x-b3-parentspanid": "437370e7b53ef921",
          "x-b3-sampled": "1",
          "x-b3-spanid": "760b19a8eafc9a9e",
          "x-b3-traceid": "005443ba0aac0f28437370e7b53ef921",
          "x-email": "",
          "x-envoy-external-address": "10.44.5.13",
          "x-envoy-original-path": "/model/wine-clone/api/model/info",
          "x-forwarded-for": "10.4.0.29,10.44.5.13, 127.0.0.1",
          "x-forwarded-host": "gke01.dev.odahu.org",
          "x-forwarded-port": "443",
          "x-forwarded-proto": "http",
          "x-jwt": "",
          "x-odahu-required-role": "test",
          "x-real-ip": "10.4.0.29",
          "x-request-id": "46b9c695-c0e5-98e8-b099-5bd882e0e910",
          "x-scheme": "https",
          "x-user": "66d42273-4b32-453e-97f5-8865d4f1a444"
        },
        "host": "10.44.9.3:8012",
        "id": "10169812507309876470",
        "method": "POST",
        "path": "/api/model/invoke",
        "protocol": "HTTP/1.1"
      }
    },
    "source": {
      "address": {
        "Address": {
          "SocketAddress": {
            "PortSpecifier": {
              "PortValue": 49894
            },
            "address": "10.44.7.177"
          }
        }
      }
    }
  }
```

3. It leads to a situation where we not able to write different policies for different pods that have the same URL paths on which we rely on

#### Proposed solution:

1. Migrate to OPA sidecars https://github.com/open-policy-agent/opa-envoy-plugin/tree/master/examples/istio
2. Inject whether the policy as cm volume statically in webhook to OPA sidecar OR inject https://github.com/open-policy-agent/kube-mgmt that will dynamically fetch policy from appropriate CM for each pod

### PR DESCRIPTION
**Pull request by vlad-tokarev**
closes https://github.com/odahu/odahu-flow/issues/408

### CODE REVIEW

File: packages/operator/controllers/modeldeployment_controller.go [hunk 1]
```diff
@@ -633,6 +634,48 @@ func (r *ModelDeploymentReconciler) cleanupOldRevisions(
...
+func (r *ModelDeploymentReconciler) reconcilePolicyCM(log logr.Logger,
+	modelDeploymentCR *odahuflowv1alpha1.ModelDeployment) error {
+
+	rn := modelDeploymentCR.Spec.RoleName
+	if rn == nil {
+		log.Info(".Spec.RoleName is nil. Skip creating custom policies for model")
+		return nil
+	}
+
+	policies, err := deployment.ReadDefaultPoliciesAndRender(*rn)
```
Comment (keshamin) [inline packages/operator/controllers/modeldeployment_controller.go:650]: I don't know how lightweight or heavy this operations is (not familiar with `bindata`), but maybe we could optimize it a bit if we check for existing config map **before** we build new one from templates. Anyway this is really minor and up to you.

**Review by keshamin — APPROVED** [review]

**Review by karbyshevds — CHANGES_REQUESTED** [review]
Please fix formatting for *.rego files.
examples:
[1](https://github.com/odahu/odahu-flow/pull/413/files#diff-8660ab2ce129e3421b0cdcca3f5e3b91ced90e585b05d634c24ef7428b3d5ff4R28)
[2](https://github.com/odahu/odahu-flow/pull/413/files#diff-8660ab2ce129e3421b0cdcca3f5e3b91ced90e585b05d634c24ef7428b3d5ff4R10)
[3](https://github.com/odahu/odahu-flow/pull/413/files#diff-40b4c43eb17639340e9d40ec9352edc0eeaf4aea2dcf70c13877430ee61b68e1R5)

**Comment by vlad-tokarev** [comment]
> Please fix formatting for *.rego files.
> examples:
> [1](https://github.com/odahu/odahu-flow/pull/413/files#diff-8660ab2ce129e3421b0cdcca3f5e3b91ced90e585b05d634c24ef7428b3d5ff4R28)
> [2](https://github.com/odahu/odahu-flow/pull/413/files#diff-8660ab2ce129e3421b0cdcca3f5e3b91ced90e585b05d634c24ef7428b3d5ff4R10)
> [3](https://github.com/odahu/odahu-flow/pull/413/files#diff-40b4c43eb17639340e9d40ec9352edc0eeaf4aea2dcf70c13877430ee61b68e1R5)

fixed. Sorry my IDE was not showing this difference for .rego files by default

File: packages/operator/controllers/modeldeployment_controller.go (hunk 1 above)
Comment (vlad-tokarev) [reply packages/operator/controllers/modeldeployment_controller.go:650]: This operation is very cheap (reading bytes from memory buffer) so I believe that this optimization will have almost no effect

**Review by karbyshevds — APPROVED** [review]

**Pull request merged**